src/classes/RD2_ScheduleService.cls

```diff
@@ -185,6 +193,152 @@ public without sharing class RD2_ScheduleService {
         return isNeeded;
     }
 
+
+    /***
+    * @description Calculates the dates of the next n donations
+    * @param rds Recurring Donation records
+    * @param visualizationStartDate Date
+    * @param maxDates Integer
+    * @return Map<Id, List<Installment>>
+    */
+    public List<Installment> getNextDonationsForVisualization(npe03__Recurring_Donation__c rd, Date visualizationStartDate, Integer maxDates) {
+        if (rd == null) {
+            return new List<Installment>();
+        }
+
+        return getNextDonationsForVisualization(new List<npe03__Recurring_Donation__c> {rd}, visualizationStartDate, maxDates).get(rd.Id);
+    }
+
+    /***
+    * @description Calculates the dates of the next n donations
+    * @param rd Recurring Donation record
+    * @param visualizationStartDate Date
+    * @param maxDates Integer
+    * @return List<Installment>
+    */
+    public Map<Id, List<Installment>> getNextDonationsForVisualization(List<npe03__Recurring_Donation__c> rds, Date visualizationStartDate, Integer maxDates) {
+
+        if (maxDates == null) {
+            maxDates = DEFAULT_VISUALIZATION_SCHEDULE_COUNT;
+        }
+
+        if (visualizationStartDate == null) {
+            visualizationStartDate = currentDate;
+        }
+
+        Map<Id, List<Installment>> installmentsByRDIds = new Map<Id, List<Installment>>();
+        Map<Id, List<RecurringDonationSchedule__c>> schedulesByRDIds = getExistingSchedules(rds);
+
+        for (npe03__Recurring_Donation__c rd : rds) {
+
+            if (rd == null) {
+                continue;
+            }
+
+            installmentsByRDIds.put(rd.Id, new List<Installment>());
+
+            Installment ins;
+            List<Installment> vizSchedules = new List<Installment>();
+            currentDate = visualizationStartDate;
+            Integer dateCount = 0;
+            List <RecurringDonationSchedule__c> filteredSchedules;
+
+            if (schedulesByRDIds.get(rd.Id) == null) {
+                continue;
+            }
+
+            while (dateCount < maxDates) {
```

Comment:
Could this just be `while (visSchedules.size() < numInstallments) {` this way the `dateCount` var isn't needed at all?